src/combinatorics/derangements/finite.lean

```diff
@@ -0,0 +1,117 @@
+/-
+Copyright (c) 2021 Henry Swanson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henry Swanson
+-/
+import combinatorics.derangements.basic
+import data.fintype.card
+import tactic.delta_instance
+import tactic.ring
+
+/-!
+# Derangements on fintypes
+
+This file contains lemmas that describe the cardinality of `derangements α` when `α` is a fintype.
+
+# Main definitions
+
+  - `card_derangements_invariant`: A lemma stating that the number of derangements on a type `α`
+    depends only on the cardinality of `α`.
+
+  - `num_derangements n`: The number of derangements on an n-element set, defined in a computation-
+    friendly way.
+
+  - `card_derangements_eq_num_derangements`: Proof that `num_derangements` really does compute the
+    number of derangements.
+
+  - `num_derangements_sum`: A lemma giving an expression for `num_derangements n` in terms of
+    factorials.
+-/
+
+open derangements equiv fintype
+open_locale big_operators
+
+variables {α : Type*} [decidable_eq α] [fintype α]
+
+instance : decidable_pred (derangements α) := λ _, fintype.decidable_forall_fintype
+
+instance : fintype (derangements α) := by delta_instance derangements
+
+lemma card_derangements_invariant {α β : Type*} [fintype α] [decidable_eq α]
+  [fintype β] [decidable_eq β] (h : card α = card β) :
+  card (derangements α) = card (derangements β) :=
+fintype.card_congr (equiv.derangements_congr $ equiv_of_card_eq h)
+
+lemma card_derangements_fin_add_two (n : ℕ) :
+  card (derangements (fin (n+2))) = (n+1) * card (derangements (fin n)) +
+  (n+1) * card (derangements (fin (n+1))) :=
+begin
+  -- get some basic results about the size of fin (n+1) plus or minus an element
+  have h1 : ∀ a : fin (n+1), card ({a}ᶜ : set (fin (n+1))) = card (fin n),
+  { intro a,
+    simp only [fintype.card_fin, finset.card_fin, fintype.card_of_finset, finset.filter_ne' _ a,
+      set.mem_compl_singleton_iff, finset.card_erase_of_mem (finset.mem_univ a), nat.pred_succ] },
+  have h2 : card (fin (n+2)) = card (option (fin (n+1))),
+  { simp only [card_fin, card_option] },
+  -- rewrite the LHS and substitute in our fintype-level equivalence
+  simp only [card_derangements_invariant h2,
+    card_congr (@derangements_recursion_equiv (fin (n+1)) _),
+  -- push the cardinality through the Σ and ⊕ so that we can use `card_n`
+    card_sigma, card_sum, card_derangements_invariant (h1 _), finset.sum_const, nsmul_eq_mul,
+    finset.card_fin, mul_add, nat.cast_id],
+end
+
+/-- The number of derangements on an `n`-element set. -/
```

Comment:
Maybe that's grammatically correcter? I don't know.
```suggestion
/-- The number of derangements of an `n`-element set. -/
```

Reply:
I think it's one of those situations where either works, like with "permutations of/on a set". I think I see "on" more often when it's a more structured object, e.g., "permutations on a lattice"? YMMV